src/backend/distributed/commands/index.c

```diff
@@ -590,13 +605,30 @@ PreprocessReindexStmt(Node *node, const char *reindexCommand,
 		Oid relationId = ReindexStmtFindRelationOid(reindexStatement, false);
 		MemoryContext relationContext = NULL;
 		Relation relation = NULL;
+		LOCKMODE lockmode = IsReindexWithParam_compat(reindexStatement, "concurrently") ?
+							ShareUpdateExclusiveLock : AccessExclusiveLock;
+
 		if (reindexStatement->kind == REINDEX_OBJECT_INDEX)
 		{
+			/*
+			 * Acquire global lock to prevent concurrent writes or DDL. However,
+			 * we do not bother for REINDEX CONCURRENTLY, since we'll have
+			 * to release the lock.
+			 */
+			if (!IsReindexWithParam_compat(reindexStatement, "concurrently"))
+			{
+				AcquireDistributedLockOnRelations(list_make1(reindexStatement->relation),
```

Comment:
Hmm, pg seems to acquire `ShareLock` on the table, and `AccessShare` on the index. We acquire `AccessExclusive` on the table itself.

Does that cause any side-effects?

Reply:
This is not actually true. When `reindexStatement->kind == REINDEX_OBJECT_INDEX`, `reindexStatement->relation` is the oid of the index, not the underlying table.